drivers/nutdrv_qx_masterguard.c

```diff
@@ -0,0 +1,1049 @@
+	upsdebugx(2, "setvar: illegal %s value %s", item->dfl, value);
+	return -1;
+}
+
+/* record new slave address in masterguard_next_slaveaddr; moved to masterguard_my_slaveaddr in masterguard_new_slaveaddr() after the slaveaddr-changing command finished */
+static int masterguard_set_slaveaddr(item_t *item, char *value, const size_t valuelen) {
+	char *p;
+
+	masterguard_next_slaveaddr = strtol(value, &p, 10);
+	if (*p != '\0') {
+		upsdebugx(2, "set_slaveaddr: illegal value %s", value);
+		return -1;
+	}
+	upsdebugx(3, "next slaveaddr %d", masterguard_next_slaveaddr);
+	return masterguard_setvar(item, value, valuelen);
+}
+
+
+/* variable setting answer preprocessing functions */
+
+/* set my_slaveaddr to next_slaveaddr /after/ issuing the SS command (which, itself, needs the /old/ slaveaddr) */
+static int masterguard_new_slaveaddr(item_t *item, const int len) {
+	upsdebugx(3, "saved slaveaddr %d", masterguard_next_slaveaddr);
+	masterguard_my_slaveaddr[0] = '0' + masterguard_next_slaveaddr / 10;
+	masterguard_my_slaveaddr[1] = '0' + masterguard_next_slaveaddr % 10;
+	upsdebugx(3, "new slaveaddr %s", masterguard_my_slaveaddr);
+	return len;
+}
+
+
+/* qx2nut lookup table */
+static item_t masterguard_qx2nut[] = {
+	/* static values */
+
+	/* type				flags	rw	command	answer	len	leading	value	from	to	dfl		qxflags				precmd	preans	preproc */
+	{ "device.mfr",			0,	NULL,	"",	"",	0,	'\0',	"",	0,	0,	"Masterguard",	QX_FLAG_STATIC | QX_FLAG_ABSENT,NULL,	NULL,	NULL },
+	{ "load.high",			0,	NULL,	"",	"",	0,	'\0',	"",	0,	0,	"140",		QX_FLAG_STATIC | QX_FLAG_ABSENT,NULL,	NULL,	NULL },
+	/* battery.charge.low */
+	/* battery.charge.warning */
+	{ "battery.type",		0,	NULL,	"",	"",	0,	'\0',	"",	0,	0,	"PbAc",		QX_FLAG_STATIC | QX_FLAG_ABSENT,NULL,	NULL,	NULL },
+
+
+	/* variables */
+
+	/*
+	 * > [WH\r]
+	 * < [(XX VV.VV PP.PP TTTTTTTTTTTTTTTTTTTTTTTTTTTTTT B MMM FF.FF VVV SS.SS HHH.hh GGG.gg RRR mm nn MMM NNN FF.FF FF.FF\r]
+	 *    01234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012
+	 *    0         1         2         3         4         5         6         7         8         9         0         1
+	 *    (00 10.06 03.09 A  700  + 0 Bat Pack-19        0 230 50.00 012 02.30 006.00 012.00 018 10 40 160 276 47.00 53.00
+	 */
+	/* type				flags		rw			command	answer	len	leading	value	from	to	dfl	qxflags					precmd	preans	preproc */
+	{ "ups.id",			ST_FLAG_RW,	masterguard_r_slaveaddr,"WH\r",	"",	113,	'(',	"",	1,	2,	"%.0f",	QX_FLAG_SEMI_STATIC | QX_FLAG_RANGE,	NULL,	NULL,	masterguard_slaveaddr },
+	{ "ups.firmware",		0,		NULL,			"WH\r",	"",	113,	'(',	"",	4,	8,	"%s",	QX_FLAG_STATIC,				NULL,	NULL,	NULL },
+	{ "ups.firmware.aux",		0,		NULL,			"WH\r",	"",	113,	'(',	"",	10,	14,	"%s",	QX_FLAG_STATIC,				NULL,	NULL,	NULL },
+	/* several values are deduced from the T field */
+	{ "series",			0,		NULL,			"WH\r",	"",	113,	'(',	"",	16,	16,	"%s",	QX_FLAG_STATIC | QX_FLAG_NONUT,		NULL,	NULL,	masterguard_series },
```

Comment:
@aquette @clepple : Any suggestions for standard field naming?

Reply:
`device.series`?